roachtest: cdc/tpcc-1000/rangefeed=true failed

SHA: https://github.com/cockroachdb/cockroach/commits/10f8010fa5778e740c057905e2d7664b5fd5d647

Parameters:

To repro, try:

```
# Don't forget to check out a clean suitable branch and experiment with the
# stress invocation until the desired results present themselves. For example,
# using stress instead of stressrace and passing the '-p' stressflag which
# controls concurrency.
./scripts/gceworker.sh start && ./scripts/gceworker.sh mosh
cd ~/go/src/github.com/cockroachdb/cockroach && \
stdbuf -oL -eL \
make stressrace TESTS=cdc/tpcc-1000/rangefeed=true PKG=roachtest TESTTIMEOUT=5m STRESSFLAGS='-maxtime 20m -timeout 10m' 2>&1 | tee /tmp/stress.log
```

Failed test: https://teamcity.cockroachdb.com/viewLog.html?buildId=1135549&tab=buildLog

```
The test failed on master:
	cluster.go:1226,cdc.go:612,cdc.go:124,cluster.go:1564,errgroup.go:57: /home/agent/work/.go/src/github.com/cockroachdb/cockroach/bin/roachprod run teamcity-1135549-cdc-tpcc-1000-rangefeed-true:4 -- ./workload run tpcc --warehouses=1000 --duration=120m --tolerate-errors {pgurl:1-3}  returned:
		stderr:
		
		stdout:
		0.0      0.0 newOrder
		     38s      688            0.0           22.4      0.0      0.0      0.0      0.0 orderStatus
		     38s      688            0.0          197.1      0.0      0.0      0.0      0.0 payment
		     38s      688            0.0           20.4      0.0      0.0      0.0      0.0 stockLevel
		E190213 08:12:19.560089 1 workload/cli/run.go:402  error in payment: dial tcp 10.142.0.98:26257: connect: connection refused
		     39s      909            0.0           19.5      0.0      0.0      0.0      0.0 delivery
		     39s      909            0.0          138.4      0.0      0.0      0.0      0.0 newOrder
		     39s      909            0.0           21.9      0.0      0.0      0.0      0.0 orderStatus
		     39s      909            0.0          192.1      0.0      0.0      0.0      0.0 payment
		     39s      909            0.0           19.9      0.0      0.0      0.0      0.0 stockLevel
		E190213 08:12:20.560203 1 workload/cli/run.go:402  error in payment: dial tcp 10.142.0.98:26257: connect: connection refused
		: signal: killed
	cluster.go:1585,cdc.go:212,cdc.go:367,test.go:1212: unexpected node event: 3: dead

```

This doesn't look good:

```
E190213 08:12:17.034655 425977 util/log/crash_reporting.go:202  [n3,s3,r4542/3:/Table/5{4/1/764/9-5},rangefeed] a panic has occurred!
panic: resolved timestamp 1550045524.323614914,0 equal to or above timestamp of operation {<nil> txn_id:70a904d1-3f15-4f50-b257-a064d80e453c txn_key:"\276\211\367\003=\211\210" timestamp:<wall_time:1550045519101462177 logical:2 >  <nil> <nil> <nil> {} 0} [recovered]
	panic: resolved timestamp 1550045524.323614914,0 equal to or above timestamp of operation {<nil> txn_id:70a904d1-3f15-4f50-b257-a064d80e453c txn_key:"\276\211\367\003=\211\210" timestamp:<wall_time:1550045519101462177 logical:2 >  <nil> <nil> <nil> {} 0}

goroutine 425977 [running]:
github.com/cockroachdb/cockroach/pkg/util/stop.(*Stopper).Recover(0xc000876000, 0x38e6ea0, 0xc01ae31740)
	/go/src/github.com/cockroachdb/cockroach/pkg/util/stop/stopper.go:185 +0x11f
panic(0x2c53760, 0xc033c08490)
	/usr/local/go/src/runtime/panic.go:513 +0x1b9
github.com/cockroachdb/cockroach/pkg/storage/rangefeed.(*resolvedTimestamp).assertOpAboveRTS(0xc04b095238, 0x0, 0xc0126eea80, 0x0, 0x0, 0x0, 0x0, 0x1582de2fe7ba86a1, 0x2)
	/go/src/github.com/cockroachdb/cockroach/pkg/storage/rangefeed/resolved_timestamp.go:212 +0x185
github.com/cockroachdb/cockroach/pkg/storage/rangefeed.(*resolvedTimestamp).consumeLogicalOp(0xc04b095238, 0x0, 0xc0126eea80, 0x0, 0x0, 0x0, 0x0, 0xc000244200)
	/go/src/github.com/cockroachdb/cockroach/pkg/storage/rangefeed/resolved_timestamp.go:145 +0x352
github.com/cockroachdb/cockroach/pkg/storage/rangefeed.(*resolvedTimestamp).ConsumeLogicalOp(0xc04b095238, 0x0, 0xc0126eea80, 0x0, 0x0, 0x0, 0x0, 0x1010000000000)
	/go/src/github.com/cockroachdb/cockroach/pkg/storage/rangefeed/resolved_timestamp.go:131 +0x4d
github.com/cockroachdb/cockroach/pkg/storage/rangefeed.(*Processor).consumeLogicalOps(0xc04b095180, 0x38e6de0, 0xc05016a940, 0xc04273c4b0, 0x1, 0x1)
	/go/src/github.com/cockroachdb/cockroach/pkg/storage/rangefeed/processor.go:531 +0x21b
github.com/cockroachdb/cockroach/pkg/storage/rangefeed.(*Processor).consumeEvent(0xc04b095180, 0x38e6de0, 0xc05016a940, 0xc04273c4b0, 0x1, 0x1, 0x0, 0x0, 0x0, 0x0, ...)
	/go/src/github.com/cockroachdb/cockroach/pkg/storage/rangefeed/processor.go:483 +0x251
github.com/cockroachdb/cockroach/pkg/storage/rangefeed.(*Processor).Start.func1(0x38e6de0, 0xc05016a940)
	/go/src/github.com/cockroachdb/cockroach/pkg/storage/rangefeed/processor.go:242 +0x9c9
github.com/cockroachdb/cockroach/pkg/util/stop.(*Stopper).RunWorker.func1(0xc00f60a270, 0xc000876000, 0xc01ae31650)
	/go/src/github.com/cockroachdb/cockroach/pkg/util/stop/stopper.go:200 +0xe1
created by github.com/cockroachdb/cockroach/pkg/util/stop.(*Stopper).RunWorker
	/go/src/github.com/cockroachdb/cockroach/pkg/util/stop/stopper.go:193 +0xa8
```

SHA: https://github.com/cockroachdb/cockroach/commits/d7c56dcb87c8f187e50303c8e32a64836c42515c

Parameters:

To repro, try:

```
# Don't forget to check out a clean suitable branch and experiment with the
# stress invocation until the desired results present themselves. For example,
# using stress instead of stressrace and passing the '-p' stressflag which
# controls concurrency.
./scripts/gceworker.sh start && ./scripts/gceworker.sh mosh
cd ~/go/src/github.com/cockroachdb/cockroach && \
stdbuf -oL -eL \
make stressrace TESTS=cdc/tpcc-1000/rangefeed=true PKG=roachtest TESTTIMEOUT=5m STRESSFLAGS='-maxtime 20m -timeout 10m' 2>&1 | tee /tmp/stress.log
```

Failed test: https://teamcity.cockroachdb.com/viewLog.html?buildId=1139797&tab=buildLog

```
The test failed on master:
	cluster.go:1226,cdc.go:612,cdc.go:124,cluster.go:1564,errgroup.go:57: /home/agent/work/.go/src/github.com/cockroachdb/cockroach/bin/roachprod run teamcity-1139797-cdc-tpcc-1000-rangefeed-true:4 -- ./workload run tpcc --warehouses=1000 --duration=120m --tolerate-errors {pgurl:1-3}  returned:
		stderr:
		
		stdout:
		   0.0 newOrder
		     49s      580            7.0           18.2      7.3     12.6     12.6     12.6 orderStatus
		     49s      580            0.0          196.3      0.0      0.0      0.0      0.0 payment
		     49s      580            1.0           20.2      7.9      7.9      7.9      7.9 stockLevel
		E190215 08:15:34.976066 1 workload/cli/run.go:402  error in orderStatus: dial tcp 10.142.0.195:26257: connect: connection refused
		     50s      770            0.0           19.7      0.0      0.0      0.0      0.0 delivery
		     50s      770            0.0          143.2      0.0      0.0      0.0      0.0 newOrder
		     50s      770            1.0           17.9      7.6      7.6      7.6      7.6 orderStatus
		     50s      770            0.0          192.4      0.0      0.0      0.0      0.0 payment
		     50s      770            5.0           19.9     71.3    151.0    151.0    151.0 stockLevel
		E190215 08:15:35.976058 1 workload/cli/run.go:402  error in payment: dial tcp 10.142.0.195:26257: connect: connection refused
		: signal: killed
	cluster.go:1585,cdc.go:212,cdc.go:367,test.go:1212: unexpected node event: 2: dead

```

SHA: https://github.com/cockroachdb/cockroach/commits/8e9a1e310e3e8e37f091b7ca8bd204084ad9e2e5

Parameters:

To repro, try:

```
# Don't forget to check out a clean suitable branch and experiment with the
# stress invocation until the desired results present themselves. For example,
# using stress instead of stressrace and passing the '-p' stressflag which
# controls concurrency.
./scripts/gceworker.sh start && ./scripts/gceworker.sh mosh
cd ~/go/src/github.com/cockroachdb/cockroach && \
stdbuf -oL -eL \
make stressrace TESTS=cdc/tpcc-1000/rangefeed=true PKG=roachtest TESTTIMEOUT=5m STRESSFLAGS='-maxtime 20m -timeout 10m' 2>&1 | tee /tmp/stress.log
```

Failed test: https://teamcity.cockroachdb.com/viewLog.html?buildId=1142461&tab=buildLog

```
The test failed on master:
	cluster.go:1226,cdc.go:612,cdc.go:124,cluster.go:1564,errgroup.go:57: /home/agent/work/.go/src/github.com/cockroachdb/cockroach/bin/roachprod run teamcity-1142461-cdc-tpcc-1000-rangefeed-true:4 -- ./workload run tpcc --warehouses=1000 --duration=120m --tolerate-errors {pgurl:1-3}  returned:
		stderr:
		
		stdout:
		   0.0 newOrder
		     18s     1375            0.0           24.4      0.0      0.0      0.0      0.0 orderStatus
		     18s     1375            0.0           52.3      0.0      0.0      0.0      0.0 payment
		     18s     1375            0.0           23.5      0.0      0.0      0.0      0.0 stockLevel
		E190217 08:01:43.787699 1 workload/cli/run.go:402  error in orderStatus: dial tcp 10.142.0.121:26257: connect: connection refused
		     19s     3039            0.0           17.9      0.0      0.0      0.0      0.0 delivery
		     19s     3039            0.0            0.0      0.0      0.0      0.0      0.0 newOrder
		     19s     3039            0.0           23.1      0.0      0.0      0.0      0.0 orderStatus
		     19s     3039            0.0           49.6      0.0      0.0      0.0      0.0 payment
		     19s     3039            0.0           22.3      0.0      0.0      0.0      0.0 stockLevel
		E190217 08:01:44.787742 1 workload/cli/run.go:402  error in payment: dial tcp 10.142.0.121:26257: connect: connection refused
		: signal: killed
	cluster.go:1585,cdc.go:212,cdc.go:367,test.go:1212: unexpected node event: 3: dead

```

Same sort of thing. The assertion fires in the write intent path (here and above).

https://github.com/cockroachdb/cockroach/blob/db4a8a1c6413b3a769914b011a88287eed4fd8ce/pkg/storage/rangefeed/resolved_timestamp.go#L144-L146

So it looks like it's telling us that the resolved timestamp (known to the `rts`) is above the intent that is being written. @nvanbenschoten this seems pretty severe -- could you help me figure this out?

```
190217 08:01:41.395375 59325 util/log/crash_reporting.go:202  [n2,s2,r1872/2:/Table/61/1/2{49/7/-â€¦-51/6/-â€¦},rangefeed] a panic has occurred!
panic: resolved timestamp 1550390488.242802874,0 equal to or above timestamp of operation {<nil> txn_id:696c4ebd-4db5-4e78-be80-82cb6c3b7363 txn_key:"\277\211\366\373\211\206\367\312\210" timestamp:<wall_time:1550390487301127124 logical:2 >  <nil> <nil> <nil> {} 0} [recovered]
	panic: resolved timestamp 1550390488.242802874,0 equal to or above timestamp of operation {<nil> txn_id:696c4ebd-4db5-4e78-be80-82cb6c3b7363 txn_key:"\277\211\366\373\211\206\367\312\210" timestamp:<wall_time:1550390487301127124 logical:2 >  <nil> <nil> <nil> {} 0}

goroutine 59325 [running]:
github.com/cockroachdb/cockroach/pkg/util/stop.(*Stopper).Recover(0xc0006be000, 0x3918560, 0xc02f3174a0)
	/go/src/github.com/cockroachdb/cockroach/pkg/util/stop/stopper.go:185 +0x11f
panic(0x2c77500, 0xc02444fe60)
	/usr/local/go/src/runtime/panic.go:513 +0x1b9
github.com/cockroachdb/cockroach/pkg/storage/rangefeed.(*resolvedTimestamp).assertOpAboveRTS(0xc0153d19b8, 0x0, 0xc036d7d400, 0x0, 0x0, 0x0, 0x0, 0x158417ef12ba7bd4, 0x2)
	/go/src/github.com/cockroachdb/cockroach/pkg/storage/rangefeed/resolved_timestamp.go:212 +0x185
github.com/cockroachdb/cockroach/pkg/storage/rangefeed.(*resolvedTimestamp).consumeLogicalOp(0xc0153d19b8, 0x0, 0xc036d7d400, 0x0, 0x0, 0x0, 0x0, 0x2)
	/go/src/github.com/cockroachdb/cockroach/pkg/storage/rangefeed/resolved_timestamp.go:145 +0x352
github.com/cockroachdb/cockroach/pkg/storage/rangefeed.(*resolvedTimestamp).ConsumeLogicalOp(0xc0153d19b8, 0x0, 0xc036d7d400, 0x0, 0x0, 0x0, 0x0, 0x10119485f379c)
	/go/src/github.com/cockroachdb/cockroach/pkg/storage/rangefeed/resolved_timestamp.go:131 +0x4d
github.com/cockroachdb/cockroach/pkg/storage/rangefeed.(*Processor).consumeLogicalOps(0xc0153d1900, 0x39184a0, 0xc007e4c9c0, 0xc013d7f000, 0x45, 0x80)
	/go/src/github.com/cockroachdb/cockroach/pkg/storage/rangefeed/processor.go:531 +0x21b
github.com/cockroachdb/cockroach/pkg/storage/rangefeed.(*Processor).consumeEvent(0xc0153d1900, 0x39184a0, 0xc007e4c9c0, 0xc013d7f000, 0x45, 0x80, 0x0, 0x0, 0x0, 0x0, ...)
	/go/src/github.com/cockroachdb/cockroach/pkg/storage/rangefeed/processor.go:483 +0x251
github.com/cockroachdb/cockroach/pkg/storage/rangefeed.(*Processor).Start.func1(0x39184a0, 0xc007e4c9c0)
	/go/src/github.com/cockroachdb/cockroach/pkg/storage/rangefeed/processor.go:242 +0x9c9
github.com/cockroachdb/cockroach/pkg/util/stop.(*Stopper).RunWorker.func1(0xc0143cf670, 0xc0006be000, 0xc02f317440)
	/go/src/github.com/cockroachdb/cockroach/pkg/util/stop/stopper.go:200 +0xe1
created by github.com/cockroachdb/cockroach/pkg/util/stop.(*Stopper).RunWorker
	/go/src/github.com/cockroachdb/cockroach/pkg/util/stop/stopper.go:193 +0xa8

```

--- a/pkg/storage/engine/mvcc.go
+++ b/pkg/storage/engine/mvcc.go
@@ -1266,9 +1266,6 @@ func replayTransactionalWrite(
 // that the timestamp parameter be set to hlc.Timestamp{} when writing
 // transactionally, but hlc.Timestamp{} is already used as a sentinel
 // for inline puts.)
-//
-// TODO(andrei): explore alternate function signatures that reduce
-// confusion about which timestamp applies.
 func mvccPutInternal(
 	ctx context.Context,
 	engine Writer,
@@ -1326,18 +1323,23 @@ func mvccPutInternal(
 		return err
 	}
 
+	// Determine the read and write timestamps for the write. For a
+	// non-transactional write, these will be identical. For a transactional
+	// write, we read at the transaction's original timestamp (forwarded by any
+	// refresh timestamp) but write intents at its provisional commit timestamp.
+	// See the comment on the txn.Timestamp field definition for rationale.
+	readTimestamp := timestamp
 	writeTimestamp := timestamp
 	if txn != nil {
-		txnReadTimestamp := txn.OrigTimestamp
-		txnReadTimestamp.Forward(txn.RefreshedTimestamp)
-		if txnReadTimestamp != timestamp {
+		readTimestamp = txn.OrigTimestamp
+		readTimestamp.Forward(txn.RefreshedTimestamp)
+		if readTimestamp != timestamp {
 			return errors.Errorf("mvccPutInternal: txn's read timestamp %s does not match timestamp %s",
 				txn.OrigTimestamp, timestamp)
 		}
-		// A txn writes intents at its provisional commit timestamp. See the
-		// comment on the txn.Timestamp field definition for rationale.
 		writeTimestamp = txn.Timestamp
 	}
+	timestamp = hlc.Timestamp{} // prevent accidental use below
 
 	// Determine what the logical operation is. Are we writing an intent
 	// or a value directly?
@@ -1371,7 +1373,7 @@ func mvccPutInternal(
 				// The transaction has executed at this sequence before. This is merely a
 				// replay of the transactional write. Assert that all is in order and return
 				// early.
-				return replayTransactionalWrite(ctx, engine, iter, meta, ms, key, timestamp, value, txn, buf, valueFn)
+				return replayTransactionalWrite(ctx, engine, iter, meta, ms, key, readTimestamp, value, txn, buf, valueFn)
 			}
 
 			// We're overwriting the intent that was present at this key, before we do
@@ -1383,11 +1385,11 @@ func mvccPutInternal(
 			getBuf.meta = buf.meta // initialize get metadata from what we've already read
 
 			existingVal, _, _, err := mvccGetInternal(
-				ctx, iter, metaKey, timestamp, true /* consistent */, safeValue, txn, getBuf)
+				ctx, iter, metaKey, readTimestamp, true /* consistent */, safeValue, txn, getBuf)
 			if err != nil {
 				return err
 			}
-			// Its possible that the existing value is nil if the intent on the key
+			// It's possible that the existing value is nil if the intent on the key
 			// has a lower epoch. We don't have to deal with this as a special case
 			// because in this case, the value isn't written to the intent history.
 			// Instead, the intent history is blown away completely.
@@ -1460,35 +1462,54 @@ func mvccPutInternal(
 			} else {
 				buf.newMeta.IntentHistory = nil
 			}
-		} else if !metaTimestamp.Less(timestamp) {
-			// This is the case where we're trying to write under a
-			// committed value. Obviously we can't do that, but we can
-			// increment our timestamp to one logical tick past the existing
-			// value and go on to write, but then return a write-too-old
-			// error indicating what the timestamp ended up being. This
-			// timestamp can then be used to increment the txn timestamp and
-			// be returned with the response.
-			writeTimestamp = metaTimestamp.Next()
-			maybeTooOldErr = &roachpb.WriteTooOldError{Timestamp: timestamp, ActualTimestamp: writeTimestamp}
+		} else if !metaTimestamp.Less(readTimestamp) {
+			// This is the case where we're trying to write under a committed
+			// value. Obviously we can't do that, but we can increment our
+			// timestamp to one logical tick past the existing value and go on
+			// to write, but then also return a write-too-old error indicating
+			// what the timestamp ended up being. This timestamp can then be
+			// used to increment the txn timestamp and be returned with the
+			// response.
+			//
+			// For this to work, this function needs to complete its mutation to
+			// the Writer even if it plans to return a write-too-old error. This
+			// allows logic that lives in evaluateBatch to determine what to do
+			// with the error and whether it should prevent the batch from being
+			// proposed or not.
+			//
+			// NB: even if metaTimestamp is less than writeTimestamp, we can't
+			// avoid the WriteTooOld error if metaTimestamp is equal to or
+			// greater than readTimestamp. This is because certain operations
+			// like ConditionalPuts and InitPuts avoid ever needing refreshes
+			// by ensuring that they propagate WriteTooOld errors immediately
+			// instead of allowing their transactions to continue and be retried
+			// before committing.
+			writeTimestamp.Forward(metaTimestamp.Next())
+			maybeTooOldErr = &roachpb.WriteTooOldError{
+				Timestamp: readTimestamp, ActualTimestamp: writeTimestamp,
+			}
 			// If we're in a transaction, always get the value at the orig
 			// timestamp.
 			if txn != nil {
 				if value, err = maybeGetValue(
-					ctx, iter, metaKey, value, ok, timestamp, txn, buf, valueFn); err != nil {
+					ctx, iter, metaKey, value, ok, readTimestamp, txn, buf, valueFn); err != nil {
 					return err
 				}
 			} else {
-				// Outside of a transaction, read the latest value and advance
-				// the write timestamp to the latest value's timestamp + 1. The
-				// new timestamp is returned to the caller in maybeTooOldErr.
+				// Outside of a transaction, the read timestamp advances to the
+				// the latest value's timestamp + 1 as well. The new timestamp
+				// is returned to the caller in maybeTooOldErr. Because we're
+				// outside of a transaction, we'll never actually commit this
+				// value, but that's a concern of evaluateBatch and not here.
+				readTimestamp = writeTimestamp
 				if value, err = maybeGetValue(
-					ctx, iter, metaKey, value, ok, writeTimestamp, txn, buf, valueFn); err != nil {
+					ctx, iter, metaKey, value, ok, readTimestamp, txn, buf, valueFn); err != nil {
 					return err
 				}
 			}
 		} else {
 			if value, err = maybeGetValue(
-				ctx, iter, metaKey, value, ok, timestamp, txn, buf, valueFn); err != nil {
+				ctx, iter, metaKey, value, ok, readTimestamp, txn, buf, valueFn); err != nil {
 				return err
 			}
 		}
@@ -1754,7 +1775,8 @@ func mvccInitPutUsingIter(
 	failOnTombstones bool,
 	txn *roachpb.Transaction,
 ) error {
-	err := mvccPutUsingIter(ctx, engine, iter, ms, key, timestamp, noValue, txn,
+	return mvccPutUsingIter(
+		ctx, engine, iter, ms, key, timestamp, noValue, txn,
 		func(existVal *roachpb.Value) ([]byte, error) {
 			if failOnTombstones && existVal != nil && len(existVal.RawBytes) == 0 {
 				// We found a tombstone and failOnTombstones is true: fail.
@@ -1767,9 +1789,7 @@ func mvccInitPutUsingIter(
 				}
 			}
 			return value.RawBytes, nil
-		},
-	)
-	return err
+		})
 }
 
 // MVCCMerge implements a merge operation. Merge adds integer values,

--- a/pkg/storage/engine/mvcc_test.go
+++ b/pkg/storage/engine/mvcc_test.go
@@ -3111,6 +3111,51 @@ func TestMVCCMultiplePutOldTimestamp(t *testing.T) {
 	}
 }
 
+// TestMVCCPutOldOrigTimestampNewCommitTimestamp tests a case where a
+// transactional Put occurs to the same key, but with an older original
+// timestamp than a pre-existing key. As always, this should result in a
+// WriteTooOld error. However, in this case the transaction has a larger
+// provisional commit timestamp than the pre-existing key. It should write
+// its intent at this timestamp instead of directly above the existing key.
+func TestMVCCPutOldOrigTimestampNewCommitTimestamp(t *testing.T) {
+	defer leaktest.AfterTest(t)()
+
+	ctx := context.Background()
+	engine := createTestEngine()
+	defer engine.Close()
+
+	err := MVCCPut(ctx, engine, nil, testKey1, hlc.Timestamp{WallTime: 3}, value1, nil)
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	// Perform a transactional Put with a transaction whose original timestamp is
+	// below the existing key's timestamp and whose provisional commit timestamp
+	// is above the existing key's timestamp.
+	txn := makeTxn(*txn1, hlc.Timestamp{WallTime: 1})
+	txn.Timestamp = hlc.Timestamp{WallTime: 5}
+	txn.Sequence++
+	err = MVCCPut(ctx, engine, nil, testKey1, txn.OrigTimestamp, value2, txn)
+
+	// Verify that the Put returned a WriteTooOld with the ActualTime set to the
+	// transactions provisional commit timestamp.
+	expTS := txn.Timestamp
+	if wtoErr, ok := err.(*roachpb.WriteTooOldError); !ok || wtoErr.ActualTimestamp != expTS {
+		t.Fatalf("expected WriteTooOldError with actual time = %s; got %s", expTS, wtoErr)
+	}
+
+	// Verify new value was actually written at the transaction's provisional
+	// commit timestamp.
+	value, _, err := MVCCGet(ctx, engine, testKey1, hlc.MaxTimestamp, MVCCGetOptions{Txn: txn})
+	if err != nil {
+		t.Fatal(err)
+	}
+	if value.Timestamp != expTS || !bytes.Equal(value2.RawBytes, value.RawBytes) {
+		t.Fatalf("expected timestamp=%s (got %s), value=%q (got %q)",
+			value.Timestamp, expTS, value2.RawBytes, value.RawBytes)
+	}
+}
+
 func TestMVCCAbortTxn(t *testing.T) {
 	defer leaktest.AfterTest(t)()
 

